File: Python3/Medium/WrongAnswer/additive-number_3.py
class Solution:
    def isAdditiveNumber(self, num: str) -> bool:
        
        def is_additive_number(num, i, j):
            if i > 0 and num[0] == "0":
                return False
            if j - i > 1 and num[i] == "0":
                return False
            a, b = int(num[:i]), int(num[i:j])
            while j < len(num):
                a, b = b, a + b
                if num[j:].startswith(str(b)):
                    j += len(str(b))
                else:
                    return False
            return True
        for i in range(1, len(num)):
            for j in range(i + 1, len(num)):
                if is_additive_number(num, i, j):
                    return True
        return False

File: Rust/Medium/WrongAnswer/additive-number_3.rs
impl Solution {
    pub fn is_additive_number(num: String) -> bool {
        

        let num = num.as_bytes();
        let n = num.len();
        let mut a = 0;
        let mut b = 0;
        for i in 0..n/2 {
            a = a * 10 + (num[i] - b'0') as i64;
            b = 0;
            for j in i+1..n {
                b = b * 10 + (num[j] - b'0') as i64;
                if Self::dfs(num, n, j+1, a, b) {
                    return true;
                }
                if b == 0 {
                    break;
                }
            }
            if a == 0 {
                break;
            }
        }
        false
    }
    pub fn dfs(num: &[u8], n: usize, i: usize, a: i64, b: i64) -> bool {
        if i == n {
            return true;
        }
        let c = a + b;
        let mut k = i;
        let mut x = 0;
        while k < n && x < c {
            x = x * 10 + (num[k] - b'0') as i64;
            k += 1;
        }
        if x == c {
            return Self::dfs(num, n, k, b, c);
        }
        false
    }
}